### PR TITLE
[SYCL][Doc] Host pipe design

### PR DESCRIPTION
**Pull request by rho180**

### CODE REVIEW

File: sycl/doc/design/HostPipes.md [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+          typename dataT,
+          typename propertiesT = properties<>>
+class pipe
+{ 
+  struct
+#ifdef __SYCL_DEVICE_ONLY__
+  [[__sycl_detail__::add_ir_global_variable_attributes(
+    "sycl-host-pipe",
+    nullptr
+    )]]
```
Comment (steffenlarsen) [inline sycl/doc/design/HostPipes.md:59]: The attribute has been renamed to [`add_ir_attributes_global_variable`](https://github.com/intel/llvm/blob/sycl/sycl/doc/design/CompileTimeProperties.md#properties-on-a-global-variable-type), which I now realize also need to be changed in the [`device_global` design document](https://github.com/intel/llvm/blob/sycl/sycl/doc/design/DeviceGlobal.md). I see that the document uses the new name further down, so I imagine this was just an oversight.

I assume there may be compile-time properties that need to be communicated to the backend. If that is the case, the use of `__sycl_detail__::add_ir_attributes_global_variable` attribute should expand the properties to their `meta_name` and `meta_value`, like in the [`device_global` design document](https://github.com/intel/llvm/blob/sycl/sycl/doc/design/DeviceGlobal.md).
Note that I plan on changing the way these are extracted slightly (see https://github.com/intel/llvm/pull/5959) but if you go with the same approach as in the current design document I will make the amendment if this is merged first.

File: sycl/doc/design/HostPipes.md [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+The `pipe` class uses a new C++ attribute `[[__sycl_detail__::host_pipe]]` on the 
+`pipe::__pipeType` type to identify the `static const __pipeType` member `__pipe`
+as a host pipe. Similar to `[[__sycl_detail__::device_global]]`, this will inform
+the front end to generate a `sycl-unique-id` for each `__pipe`. The `pipe` class
+also introduces the global variable attribute `sycl-host-pipe` attribute to inform the sycl-post-link tool
+to generate the SPIR-V decoration `HostAccessINTEL` for each `__pipe` using the
+`sycl-unique-id` generated. 
+
+As these attributes are only needed for the device compiler, the `#ifdef __SYCL_DEVICE_ONLY__` 
+allows the customer to ue another host compiler, even if it does not recognize these attributes.
```
Comment (steffenlarsen) [inline sycl/doc/design/HostPipes.md:43]:
```suggestion
allows the customer to use another host compiler, even if it does not recognize these attributes.
```

**Comment by rho180** [comment]
@steffenlarsen @gmlueck Reviving this review as this work which was previously put on hold is now needed.

File: sycl/doc/design/HostPipes.md [hunk 3]
```diff
@@ -0,0 +1,193 @@
+# Implementation design for "Host Pipes"
+
+This document describes the implementation design for the host pipes section
+of the DPC++ extension [SYCL_INTEL_data_flow_pipes][1]. Pipes are a FIFO construct 
+that provide links between elements of a design that are accessed through read 
+and write application programming interfaces (APIs), without the notion of a
+memory address/pointer to elements within the FIFO. A host pipe is a pipe that 
+links a device kernel with a host program.
+
+[1]: https://github.com/intel-sandbox/ip-authoring-specs/blob/main/Pipe/Spec/data_flow_pipes.asciidoc
```
Comment (steffenlarsen) [inline sycl/doc/design/HostPipes.md:10]:
```suggestion
[1]: <../extensions/supported/sycl_ext_intel_dataflow_pipes.asciidoc>
```
@gmlueck - Do you know if it is intentional that the sycl_ext_intel_dataflow_pipes extension is under supported extensions?

File: sycl/doc/design/HostPipes.md [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+they are considered implementation details of DPC++.  We do not intend to
+support them as general attributes that customer code can use.
+
+```
+template <typename name, typename dataT, typename propertiesT = properties<>>
+class pipe {/*...*/};
+
+// Partial specialization to make propertiesT visible as a parameter pack
+// of properties.
+template <typename name, typename dataT, typename ...Props>
```
Comment (steffenlarsen) [inline sycl/doc/design/HostPipes.md:54]:
```suggestion
template <typename Name, typename DataT, typename ...Props>
```
Nit just for consistency.

File: sycl/doc/design/HostPipes.md [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+Generating a unique pipe id is addressed in Open Questions below.
+
+### Changes to the DPC++ runtime
+
+Several changes are needed to the DPC++ runtime
+
+* As we noted above, the front-end generates new content in the integration
+  footer which calls the function `sycl::detail::host_pipe_map::add()`.
+  The runtime defines this function and maintains information about all the
+  device global variables in the application.  This information includes:
```
Comment (steffenlarsen) [inline sycl/doc/design/HostPipes.md:172]:
```suggestion
  host pipe variables in the application.  This information includes:
```

File: sycl/doc/design/HostPipes.md (hunk 3 above)
Comment (gmlueck) [reply sycl/doc/design/HostPipes.md:10]: Yes, this is intentional.  The dataflow pipes extension was created before we reorganized the extension documents.  My understanding is that there is production FPGA application code using that extension, so we decided to make it "supported".

**Review by steffenlarsen — APPROVED** [review]
LGTM!

File: sycl/doc/design/HostPipes.md [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+}
+
+} // namespace (unnamed)
+} // namespace sycl::detail
+```
+
+Further details on adherence to C++ rules for unconstructed objects can be found
+in the [device_global][3] design.
+
+[3]: <DeviceGlobal.md>
```
Comment (bader) [inline sycl/doc/design/HostPipes.md:161]: Reference [3] is defined at line 83. This is detected by pre-commit.
Please, make sure that pre-commit checks passes.

**Review by bader — CHANGES_REQUESTED** [review]

File: sycl/doc/design/HostPipes.md (hunk 6 above)
Comment (rho180) [reply sycl/doc/design/HostPipes.md]: Fixed.

Comment (bader) [reply sycl/doc/design/HostPipes.md]: Thanks!
Sorry for the delay.
Unfortunately, I don't see pre-commit results for this PR. Could you sync your branch with the tip of the `sycl` branch to trigger a new pre-commit validation in CI system, please?

File: sycl/doc/design/HostPipes.md [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+}
+
+} // namespace (unnamed)
+} // namespace sycl::detail
+```
+
+Further details on adherence to C++ rules for unconstructed objects can be found
+in the [device_global][2] design.
+
+Generating a unique pipe id is addressed in Open Questions below.
```
Comment (bader) [inline sycl/doc/design/HostPipes.md:161]: I can't find "Open Questions" section below.

**Review by bader — APPROVED** [review]

**Pull request merged**